---
title: "Figure3"
output: html_document
---


```{r global_options, echo=FALSE}
  short=TRUE #if short==TRUE, do not echo code chunks
  debug=FALSE
```


```{r, echo=FALSE, message=FALSE}
library(rtracklayer)
  library(reshape2)
  library(grid)
  library(ggplot2)
library(GenomicRanges)

```


```{r}
ch.three = import.chain("resources/mm9ToMm10.over.chain")
```



```{r, echo=FALSE, message=FALSE}
library(Gviz)
library(GenomicFeatures)
library(biomaRt)
library(ggplot2)
library(GenomicInteractions)
library(rtracklayer)
```

```{r, message=FALSE}
mm10.ens.db = makeTxDbFromGFF("~/Downloads/Mus_musculus.GRCm38.102.chr.gtf")
ens.genes = genes(mm10.ens.db)
mouse = useEnsembl(biomart="ENSEMBL_MART_ENSEMBL", dataset="mmusculus_gene_ensembl", host = "asia.ensembl.org",  version="102")

ens.genes = ens.genes[ seqnames(ens.genes) %in% c(paste("chr", 1:19, sep=""), "chrX", "chrY")]
bm.annotations = getBM(attributes=c("ensembl_gene_id", "gene_biotype", "mgi_symbol"), mart=mouse, filters="ensembl_gene_id", values=ens.genes$gene_id, uniqueRows=TRUE)
ens.genes$mgi_symbol = bm.annotations$mgi_symbol[ match(ens.genes$gene_id, bm.annotations$ensembl_gene_id) ]
ens.genes$gene_biotype = bm.annotations$gene_biotype[ match(ens.genes$gene_id, bm.annotations$ensembl_gene_id) ]
ens.genes$description = bm.annotations$description[ match(ens.genes$gene_id, bm.annotations$ensembl_gene_id) ]
ens.genes$status = bm.annotations$status[ match(ens.genes$gene_id, bm.annotations$ensembl_gene_id) ]
ens.genes = ens.genes[ seqnames(ens.genes) %in% c(paste("chr", 1:22, sep=""), "chrX", "chrY")]
ens.genes = ens.genes[ ens.genes$gene_biotype %in% c( "protein_coding", "miRNA" ) ]
# TODO DO I NEED TO REMOVE THESE IN THE NEW VERSION?
ens.genes = ens.genes[ ens.genes$mgi_symbol != "Gm10290"] # this is a misannotated pseudogene
ens.genes = ens.genes[ ens.genes$mgi_symbol != "Gm3272"] # this is a misannotated pseudogene
ens.genes = ens.genes[ ens.genes$mgi_symbol != "4933436C20Rik"] # this is a misannotated pseudogenes
ens.genes = ens.genes[ ens.genes$gene_id != "ENSMUSG00000093049"]
```



```{r}
x = read.delim("./resources/ST1_embryo_mouse_irx3.bed") 
colnames(x)[1]='chr'
x.gr = GRanges(x$chr, IRanges(x$Start, x$End), log.p.value=x$X.log.p.value., p.value=exp(-1 * x$X.log.p.value.))
results = liftOver(x.gr, ch.three)
irx3_embryo.gr = do.call(c, results)


x = read.delim("./resources/ST1_embryo_mouse_fto.bed") 
colnames(x)[1]='chr'
x.gr = GRanges(x$chr, IRanges(x$Start, x$End), log.p.value=x$X.LOG.p.value., p.value=exp(-1 * x$X.LOG.p.value.))
results = liftOver(x.gr, ch.three)
fto_embryo.gr = do.call(c, results)

x = read.delim("./resources/ST1_adult_mouse_fto.bed") 
colnames(x)[1]='chr'
x.gr = GRanges(x$chr, IRanges(x$Start, x$End), log.p.value=x$X.LOG.p.value., p.value=exp(-1 * x$X.LOG.p.value.))
results = liftOver(x.gr, ch.three)
fto_adult.gr = do.call(c, results)


x = read.delim("./resources/ST1_adult_mouse_irx3.bed") 
colnames(x)[1]='chr'
x.gr = GRanges(x$chr, IRanges(x$Start, x$End), log.p.value=x$X.LOG.p.value., p.value=exp(-1 * x$X.LOG.p.value.))
results = liftOver(x.gr, ch.three)
irx3_adult.gr = do.call(c, results)
```



```{r figure3}

genome = "mm10"

chr="chr8"
start = 90500000
end = 93000000
axisTrack <- GenomeAxisTrack()

geneTrack = AnnotationTrack(ens.genes, genome=genome, name="Ensembl Genes", id=ifelse(ens.genes$mgi_symbol=="", " ", ens.genes$mgi_symbol))
displayPars(geneTrack) <- list(showFeatureId=TRUE, fill = "deepskyblue", fontcolor.feature = "black", fontsize=9, just.group="below")


irx3_embryo.gi = GenomicInteractions(anchor1=rep(GRanges("chr8", IRanges(start=91802067, width=10)), length(irx3_embryo.gr)), anchor2 = irx3_embryo.gr, counts=1, p.value=irx3_embryo.gr$p.value, fdr=p.adjust(irx3_embryo.gr$p.value, method="fdr"))

irx3_embryo.gi = irx3_embryo.gi[irx3_embryo.gi$fdr < 0.01]                                     
intTrack_irx3_embryo = InteractionTrack(irx3_embryo.gi)
displayPars(intTrack_irx3_embryo) = list(col.interactions="red", 
                                      col.anchors.fill ="blue",
                                      anchor.height=0.0001,
                                      plot.anchors=FALSE,
                                      interaction.dimension="height", 
                                      interaction.measure ="fdr",
                                      interaction.dimension.transform = "log")

fto_embryo.gi = GenomicInteractions(anchor1=rep(GRanges("chr8", IRanges(start=91313525, width=10)), length(fto_embryo.gr)), anchor2 = fto_embryo.gr, counts=1, p.value=fto_embryo.gr$p.value, fdr=p.adjust(fto_embryo.gr$p.value, method="fdr"))
                                      
fto_embryo.gi = fto_embryo.gi[fto_embryo.gi$fdr < 0.01]                                                           
intTrack_fto_embryo = InteractionTrack(fto_embryo.gi)
displayPars(intTrack_fto_embryo) = list(col.interactions="red", 
                                      col.anchors.fill ="blue",
                                      anchor.height=0.0001,
                                      plot.anchors=FALSE,
                                      interaction.dimension="height", 
                                      interaction.measure ="fdr",
                                      interaction.dimension.transform = "log")

fto_adult.gi = GenomicInteractions(anchor1=rep(GRanges("chr8", IRanges(start=91313525, width=10)), length(fto_adult.gr)), anchor2 = fto_adult.gr, counts=1, p.value=fto_adult.gr$p.value, fdr=p.adjust(fto_adult.gr$p.value, method="fdr"))
                                      
fto_adult.gi = fto_adult.gi[fto_adult.gi$fdr < 0.01] 

intTrack_fto_adult = InteractionTrack(fto_adult.gi)
displayPars(intTrack_fto_adult) = list(col.interactions="red", 
                                      col.anchors.fill ="blue",
                                      anchor.height=0.0001,
                                      plot.anchors=FALSE,
                                      interaction.dimension="height", 
                                      interaction.measure ="fdr",
                                      interaction.dimension.transform = "log")

irx3_adult.gi = GenomicInteractions(anchor1=rep(GRanges("chr8", IRanges(start=91802067, width=10)), length(irx3_adult.gr)), anchor2 = irx3_adult.gr, counts=1, p.value=irx3_adult.gr$p.value, fdr=p.adjust(irx3_adult.gr$p.value, method="fdr"))
                                      
irx3_adult.gi = irx3_adult.gi[irx3_adult.gi$fdr < 0.01]                                                           
intTrack_irx3_adult = InteractionTrack(irx3_adult.gi)
displayPars(intTrack_irx3_embryo) = list(col.interactions="red", 
                                      col.anchors.fill ="blue",
                                      anchor.height=0.0001,
                                      plot.anchors=FALSE,
                                      interaction.dimension="height", 
                                      interaction.measure ="fdr",
                                      interaction.dimension.transform = "log")

galGal4.70.50.density.bw = import("./resources/HCNE_density_mm10_galGal4_70pc_50col.bigWig") # obtained from ancora

density_chicken <- DataTrack(range = galGal4.70.50.density.bw, genome = genome, type = "horizon", chromosome = chr, name = "70%/50bp",  fill.horizon = c("#B41414", "#E03231", "#F7A99C", "yellow", "orange", "red"), 
                                      separator=1, lwd=0.5, ylim=c(0,12),
                                      horizon.origin=0, horizon.scale=6)


cnes = import("./resources/HCNE_mm10_galGal4_70pc_50col.bed") # obtained from ancora 

cneTrack = AnnotationTrack(reduce(subsetByOverlaps(cnes, GRanges(chr, IRanges(start - 500000, end + 500000)))), stacking="squish", genome=genome, chromosome=chr)

topdom_tads = read.delim("./resources/mNPC_GSE96107_f1_40kb_d0_chr8_w15.domain", sep="\t")
topdom_tads = topdom_tads[topdom_tads$tag != 'gap', ]
topdom_tads.gr = GRanges(topdom_tads$chr, IRanges(topdom_tads$from.coord+1, topdom_tads$to.coord))

strand(topdom_tads.gr) = "*"
tad_track = AnnotationTrack(topdom_tads.gr)
displayPars(tad_track) = list(col="black", fill="gray50")

tss = resize(ens.genes, fix="start", width=1)[seqnames(ens.genes) == chr]
tss = ens.genes[seqnames(ens.genes) == chr]
all.bp = GRanges(chr, IRanges(seq(start, end), width=1))
n = nearest(all.bp, tss, ignore.strand=TRUE)

all.bp$nearest = ifelse(tss$mgi_symbol[n]=="",
                        tss$gene_id[n], 
                        tss$mgi_symbol[n])

nearest.segments = split(all.bp, all.bp$nearest)
for(i in names(nearest.segments)){
  print(i)
  x = reduce(nearest.segments[[i]])
  x$nearest = i
  nearest.segments[[i]] = x
  
}
nearest.segments =   do.call(c, unname(nearest.segments))

nearestTrack = AnnotationTrack(nearest.segments, genome=genome, name="NEAREST", id=nearest.segments$nearest)

loc1 = start(resize(ens.genes, fix="start", width=1)[nearest(reduce(c(anchorTwo(fto_adult.gi), anchorTwo(irx3_embryo.gi), anchorTwo(irx3_adult.gi))), ens.genes, ignore.strand=TRUE)])
loc2 = start(resize(reduce(c(anchorTwo(fto_adult.gi), anchorTwo(irx3_embryo.gi),anchorTwo(irx3_adult.gi))), fix="center", width=1))

foo = GRanges(chr, IRanges(ifelse(loc1<loc2, loc1, loc2), ifelse(loc1<loc2, loc2, loc1)))
              
nearestLineTrackAll = AnnotationTrack(foo, genome=genome, name="NEAREST")

snps = GRanges(chr, IRanges(c(91374372,91379994,91390473,91384146), width=1), id =c("rs1421085", "rs1477196", "rs9939609", "rs17817449"))

snpTrack = AnnotationTrack(snps, genome=genome, name="SNPs", id = snps$id)
displayPars(snpTrack) <- list(showFeatureId=TRUE, fill = "deepskyblue", fontcolor.feature = "black", fontsize=9, just.group="below")

plotTracks(list(axisTrack,
                tad_track,
                density_chicken,
                cneTrack,
                intTrack_fto_embryo,
                intTrack_fto_adult,
                intTrack_irx3_embryo,
                intTrack_irx3_adult,
                nearestLineTrackAll,
                snpTrack,
                nearestTrack,
                geneTrack), chromosome=chr, from=start, to=end)



```



```{r}
sessionInfo()
```